### PR TITLE
[fix](stacktrace) Speed up stack trace

### PR DESCRIPTION
**Pull request by xinyiZzz**
## Proposed changes

Introduce libunwind get stack trace, cost is negligible and has line numbers.
use StackTraceCache, PHDRCache speed up, is customizable and has some optimizations.
Other stack trace tools remain: glog, boost, glibc, in case for need.

TODO:
  1. currently support `linux __x86_64__, __arm__, __powerpc__`, not supported `__FreeBSD__`, `__APPLE__`
      Note: `__arm__, __powerpc__` not been verified
  2. Support signal handle
  3. libunwid support unw_backtrace for jemalloc
  4. Use of undefined compile option USE_MUSL for later

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xinyiZzz** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 52.2 seconds
 stream load tsv:          504 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.7 seconds inserted 10000000 Rows, about 348K ops/s
 storage size: 17168331222 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230712174754_clickbench_pr_177075.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.84 seconds
 stream load tsv:          502 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162022146 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230712174754_clickbench_pr_177072.html

**Comment by xinyiZzz** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.26 seconds
 stream load tsv:          453 seconds loaded 74807831229 Bytes, about 157 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          25.6 seconds inserted 10000000 Rows, about 390K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230712101744_clickbench_pr_177071.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 54.53 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          30.5 seconds inserted 10000000 Rows, about 327K ops/s
 storage size: 17168000469 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230712184206_clickbench_pr_177178.html

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/vec/common/allocator.h [hunk 1]
```diff
@@ -69,6 +70,8 @@ static constexpr size_t MALLOC_MIN_ALIGNMENT = 8;
 // is always a multiple of sixteen. (https://www.gnu.org/software/libc/manual/html_node/Aligned-Memory-Blocks.html)
 static constexpr int ALLOCATOR_ALIGNMENT_16 = 16;
 
+inline thread_local bool in_allocator = false;
```
Comment (yiguolei) [inline be/src/vec/common/allocator.h:73]: 不能使用这种变量，因为我们不知道是否再bthread 里，比如exchange sink operator 那里。

File: be/src/util/mem_info.cpp [hunk 2]
```diff
@@ -86,21 +86,26 @@ void MemInfo::refresh_allocator_mem() {
 
     // https://jemalloc.net/jemalloc.3.html
     // https://www.bookstack.cn/read/aliyun-rds-core/4a0cdf677f62feb3.md
-    _s_allocator_cache_mem = get_je_all_arena_metrics("tcache_bytes") +
-                             get_je_metrics("stats.metadata") +
-                             get_je_all_arena_metrics("pdirty") * get_page_size();
-    _s_allocator_cache_mem_str =
-            PrettyPrinter::print(static_cast<uint64_t>(_s_allocator_cache_mem), TUnit::BYTES);
-    _s_virtual_memory_used = get_je_metrics("stats.mapped");
+    _s_allocator_cache_mem.store(get_je_all_arena_metrics("tcache_bytes") +
```
Comment (yiguolei) [inline be/src/util/mem_info.cpp:89]: 这些代码跟stacktrace 是不是没关系，先不要在这里提交了

File: be/src/runtime/memory/thread_mem_tracker_mgr.h [hunk 3]
```diff
@@ -173,7 +174,7 @@ inline void ThreadMemTrackerMgr::consume(int64_t size) {
     }
     // Large memory alloc should use allocator.h
     // Direct malloc or new large memory, unable to catch std::bad_alloc, BE may OOM.
-    if (size > 1024l * 1024 * 1024 && !doris::config::disable_memory_gc) { // 1G
+    if (size > 1024l * 1024 * 1024 && !doris::config::disable_memory_gc && !in_allocator) { // 1G
```
Comment (yiguolei) [inline be/src/runtime/memory/thread_mem_tracker_mgr.h:177]: 实在不行，就别用in_allocator 来控制了，多打几次stacktrace 也没关系

File: be/CMakeLists.txt [hunk 4]
```diff
@@ -251,6 +254,9 @@ set_target_properties(jemalloc PROPERTIES IMPORTED_LOCATION ${THIRDPARTY_DIR}/li
 add_library(jemalloc_arrow STATIC IMPORTED)
 set_target_properties(jemalloc_arrow PROPERTIES IMPORTED_LOCATION ${THIRDPARTY_DIR}/lib64/libjemalloc.a)
 
+add_library(libunwind STATIC IMPORTED)
```
Comment (yiguolei) [inline be/CMakeLists.txt:257]: 这里可能有一个风险，我们还是得弄明白，之前的pr 把libunwind 删除的原因，到底是啥冲突

File: be/src/util/stack_util.cpp [hunk 5]
```diff
@@ -25,15 +35,54 @@ void DumpStackTraceToString(std::string* stacktrace);
 
 namespace doris {
 
-// `boost::stacktrace::stacktrace()` has memory leak, so use the glog internal func to print stacktrace.
-// The reason for the boost::stacktrace memory leak is that a state is saved in the thread local of each
-// thread but is not actively released. Refer to:
-// https://github.com/boostorg/stacktrace/issues/118
-// https://github.com/boostorg/stacktrace/issues/111
 std::string get_stack_trace() {
+#ifdef ENABLE_STACKTRACE
```
Comment (yiguolei) [inline be/src/util/stack_util.cpp:39]: 这个macro 逻辑感觉是错的。 假如我们没有定义这个macro，感觉if else 是错的

File: be/src/vec/common/allocator.h (hunk 1 above)
Comment (yiguolei) [reply be/src/vec/common/allocator.h:73]: 我们得想别的办法来解决这种问题

File: be/src/util/mem_info.cpp [hunk 6]
```diff
@@ -60,14 +60,14 @@ std::string MemInfo::_s_mem_limit_str = "";
 int64_t MemInfo::_s_soft_mem_limit = -1;
 std::string MemInfo::_s_soft_mem_limit_str = "";
 
-int64_t MemInfo::_s_allocator_cache_mem = 0;
+std::atomic<int64_t> MemInfo::_s_allocator_cache_mem = 0;
```
Comment (yiguolei) [inline be/src/util/mem_info.cpp:63]: 不要带这些代码，我们这次的pr 只保留stacktrace的，这样我们才能往1.2-lts pick，太多别的代码，就不敢pick了。

File: be/src/io/fs/local_file_reader.h [hunk 7]
```diff
@@ -49,6 +49,8 @@ class LocalFileReader final : public FileReader {
 
     FileSystemSPtr fs() const override { return _fs; }
 
+    void* read_at_by_mmap();
```
Comment (yiguolei) [inline be/src/io/fs/local_file_reader.h:52]: 我不确定1.2 是否有local file reader 这个代码，如果pick 到1.2 可能不好搞。 如果有可能的话，我觉得我们直接用std::filesystem 来读取文件吧

File: be/src/vec/common/allocator.h (hunk 1 above)
Comment (xinyiZzz) [reply be/src/vec/common/allocator.h:73]: 我先放到 thread context 里，那里处理了bthread切换

File: be/src/util/mem_info.cpp (hunk 2 above)
Comment (xinyiZzz) [reply be/src/util/mem_info.cpp:89]: 好，我把和 stacktrace 不想关的先移出去

File: be/src/runtime/memory/thread_mem_tracker_mgr.h (hunk 3 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/thread_mem_tracker_mgr.h:177]: 同上，放到thread context应该没问题

File: be/CMakeLists.txt (hunk 4 above)
Comment (xinyiZzz) [reply be/CMakeLists.txt:257]: 弄明白了，原因见这两个pr，当系统包含 libunwind 时，glog会自动寻找 libunwind 依赖项加载，如果通过yum将libunwind安装在/usr/lib下会导致编译错误
https://github.com/apache/doris/pull/7866
https://github.com/StarRocks/starrocks/pull/2093

所以这个pr没有直接 -lunwind 用系统依赖，在 thridparty 中增加 libunwind 手动加载，避免可能出现的类似问题

File: be/src/util/stack_util.cpp (hunk 5 above)
Comment (xinyiZzz) [reply be/src/util/stack_util.cpp:39]: 没问题，没有定义的 macro 默认就是false，直接 `return "no enable stack"`，
代码里有个 `USE_MUSL` 也没定义，默认就是false，为了避免以后用到的时候忘了就留着了

File: be/src/util/mem_info.cpp (hunk 6 above)
Comment (xinyiZzz) [reply be/src/util/mem_info.cpp:63]: 好的

File: be/src/io/fs/local_file_reader.h (hunk 7 above)
Comment (xinyiZzz) [reply be/src/io/fs/local_file_reader.h:52]: 1.2 可能没有 local file reader ，相关代码是存算分离加的，好，我单独读吧

File: be/src/common/status.h [hunk 8]
```diff
@@ -361,7 +363,7 @@ class Status {
 #ifdef ENABLE_STACKTRACE
         if (stacktrace && capture_stacktrace(code)) {
             status._err_msg->_stack = get_stack_trace();
-            LOG(WARNING) << "meet error status: " << status;
+            // LOG(WARNING) << "meet error status: " << status; // print only where needed, may print too many stacks.
```
Comment (yiguolei) [inline be/src/common/status.h:366]: 这个不要删，这个是因为我们status 可能返回了，不打印，所以xl 加的。

Comment (xinyiZzz) [reply be/src/common/status.h]: done

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 52.39 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17166589504 Bytes

**Comment by xinyiZzz** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.24 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17168720785 Bytes

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by adonis0147 — APPROVED** [review]
LGTM

**Pull request merged**